### DISCUSSION TITLE
Migration from 1.3 to 1.4 replacing the removed `Query._entities`

### DISCUSSION THREAD

**Discussion by sphuber** [sqlalchemy/sqlalchemy#8564] [Usage Questions]
We are migrating a code base from SqlAlchemy v1.3 to v1.4 (and ideally to v2.0 afterwards). I am having trouble with one particular part of the code that was accessing the private attribute `_entities` of the `Query` class, which has disappeared in v1.4.

Essentially, the code provides a high level query builder with its own DSL. In the implementation, it will always start as follows:
```python
query = Session().query(SomeModel.id)
```
and project the primary key of the model just as a starting point.
Then it will loop over the joined models and extend the query.

In the DSL it is possible that the first model is not actually projected, and so the code used to do the following hack to remove the first entity added to the query, because it was just used as a bootstrap of sorts:
```python
query._entities.pop(0)  # Remove the first projected model
```
However, the `Query._entities` attribute was removed in v1.4 and I am not sure how to replicate this behavior with the new API. Any suggestions would be highly welcome. Thanks a lot in advance.

**Comment by zzzeek** [comment]
Hi -

the use case for accessing `Query._entities` in a read only fashion was provided using [Query.column_descriptions](https://docs.sqlalchemy.org/en/14/orm/query.html?highlight=query+column_descriptions#sqlalchemy.orm.Query.column_descriptions) that's been there for many years.

for the use case of popping from `_entities`, you have the option to re-establish new entities using [Query.with_entities](https://docs.sqlalchemy.org/en/14/orm/query.html?highlight=query+with_entities#sqlalchemy.orm.Query.with_entities).      You can combine that with column_descriptions to get the information about the remaining entities

    query.with_entities(*[desc['entity'] for desc in query.column_descriptions[1:]])
